presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -223,7 +224,7 @@ public PlanWithProperties visitOutput(OutputNode node, Context context)
         {
             PlanWithProperties child = planChild(node, context.withPreferredProperties(PreferredProperties.undistributed()));
 
-            if (!child.getProperties().isSingleNode()) {
+            if (!child.getProperties().isSingleNode() && isAddFinalGather(session)) {
```

Comment:
You should probably have a test covering plan distribution. Maybe something like `TestUnion` does.

Also, this change impacts JDBC & CLI, since output is not single node. There should be a test covering their correctness regardless of the `add-final-gather` switch value.

Reply:
Client won't see any difference. This change only impacts how many workers coordinator pulls the data from.
Regarding the test, I'm actually looking for a good way to add a unit test for a plan distribution. I'll check TestUnion out.